File: Sliding Window/643. Maximum Average Subarray I.py
class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        curr_sum = 0

        for i in range(k):
            curr_sum += nums[i]
        
        max_avg = curr_sum / k

        for i in range(k, len(nums)):
            curr_sum -= nums[i - k]
            curr_sum += nums[i]

            curr_avg = curr_sum / k
            max_avg = max(curr_avg, max_avg)
        
        return max_avg

# Time: O(n)
# Space: O(1)

File: Sliding Window/643. Maximum Average Subarray I/643. Maximum Average Subarray I.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findMaxAverage = function (nums, k) {
  let current_sum = nums.slice(0, k).reduce((a, c) => a + c, 0);
  let max_avg = current_sum / k;

  for (let i = k; i < nums.length; i++) {
    current_sum += nums[i] - nums[i - k];
    max_avg = Math.max(max_avg, current_sum / k);
  }

  return max_avg;
};

// Time: O(n)
// Space: O(1)
